### PR TITLE
Fix Bug 1480508: Add search shortcuts for new and existing users

### PR DESCRIPTION
**Pull request by AdamHillier**
WIP, based off Andrei's PR. Will need to rebase when his lands and add more tests.

### CODE REVIEW

**Comment by k88hudson** [comment]
Ok, after some manual testing this is working well in the default case, however for organic top sites, google/amazon are getting pinned to the first and second positions in all cases whereas they should be pinned to their organic positions. Looking at the code now

File: lib/TopSitesFeed.jsm [hunk 1]
```diff
@@ -70,10 +71,13 @@ this.TopSitesFeed = class TopSitesFeed {
     PageThumbs.addExpirationFilter(this);
   }
 
-  init() {
+  async init() {
     // If the feed was previously disabled PREFS_INITIAL_VALUES was never received
     this.refreshDefaults(this.store.getState().Prefs.values[DEFAULT_SITES_PREF]);
     this._storage = this.store.dbStorage.getDbTable("sectionPrefs");
+    if (this.store.getState().Prefs.values[SEARCH_SHORTCUTS_EXPERIMENT]) {
```
Comment (k88hudson) [inline lib/TopSitesFeed.jsm]: I don't think it's sufficient to only run this on `init`, this also need needs to be run when the pref updates etc. (e.g. when shield sets the experiment pref, particularly if we decide to ship to release with this default-off)

Perhaps this should be checked on every refresh i.e. in `getLinksWithDefaults`?

**Comment by AdamHillier** [comment]
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1481680

File: lib/TopSitesFeed.jsm [hunk 2]
```diff
@@ -146,6 +150,65 @@ this.TopSitesFeed = class TopSitesFeed {
...
+        Array(numberOfSlots - plainPinnedSites.length).fill(null)
+      );
+
+      await new Promise(resolve => Services.search.init(resolve));
+
+      const tryToInsertSearchShortcut = shortcut => {
+        const nextAvailable = pinnedSites.indexOf(null);
+        // Only add a search shortcut if the site isn't already pinned, we
+        // haven't previously inserted it, there's space to pin it, and the
+        // search engine is available in Firefox
```
Comment (k88hudson) [inline lib/TopSitesFeed.jsm]: :+1:

**Pull request merged**